java-lib/src/main/java/com/yammer/metrics/core/WavefrontHistogram.java

```diff
@@ -91,18 +97,26 @@ public void update(int value) {
   }
 
   /**
-   * Helper to retrieve the current bin. Assumes proper synchronization.
+   * Helper to retrieve the current bin. Will be invoked per thread.
    */
   private MinuteBin getCurrent() {
-    long minMillis = minMillis();
-
-    if (bins.isEmpty() || bins.getLast().minMillis != minMillis) {
-      bins.add(new MinuteBin(minMillis));
-      if (bins.size() > MAX_BINS) {
-        bins.removeFirst();
+    long currMinMillis = minMillis();
+
+    long key = Thread.currentThread().getId();
+    LinkedList<MinuteBin> bins = perThreadHistogramBins.getOrDefault(key, new LinkedList<>());
+    perThreadHistogramBins.putIfAbsent(key, bins);
```

Comment:
you need to check the output of this since you should be using the return value if it already exists (this doesn't happen since the thread id is unique to the thread I understand)

the typical paradigm for concurrent map is:
```
V value = map.get(key)
if (value == null) {
  // create a new value (only in high contention would this be created but discarded)
  value = new ...
  V existing = map.putIfAbsent(key, value);
  if (existing != null) value = existing;
}
```